keras/src/trainers/compile_utils.py

```diff
@@ -454,218 +459,265 @@ def variables(self):
             vars.extend(m.variables)
         return vars
 
-    def build(self, y_true, y_pred):
-        loss = self._user_loss
-        loss_weights = self._user_loss_weights
-        output_names = self._get_y_pred_output_names(y_pred)
-        inferred_output_names = output_names or self.output_names
-
-        if is_function_like(loss) and tree.is_nested(y_pred):
-            # The model has multiple outputs but only one loss fn
-            # was provided. Broadcast loss to all outputs.
-            loss = tree.map_structure(lambda x: loss, y_pred)
+    def _build_nested(self, y_true, y_pred, loss, output_names, current_path):
+        flat_y_pred = tree.flatten(y_pred)
+        if not tree.is_nested(loss):
+            _loss = loss.loss
+            if _loss is None:
+                return
+            loss_weight = loss.weight
+            resolved_loss = get_loss(_loss, y_true, y_pred)
+            name_path = current_path
+            if not tree.is_nested(output_names):
+                if output_names is not None:
+                    output_name = output_names
+                else:
+                    output_name = resolved_loss.name
+                if len(name_path) == 0:
+                    name_path = (output_name,)
+                elif isinstance(name_path[-1], int):
+                    name_path = name_path[:-1] + (output_name,)
+            name = "/".join([str(path) for path in name_path])
+            if name == "":
+                if isinstance(output_names, dict):
+                    flat_output_names = list(output_names.keys())
+                else:
+                    flat_output_names = tree.flatten(output_names)
 
-        # Check and filter the keys.
-        if isinstance(loss, dict):
-            if inferred_output_names is None:
-                raise ValueError(
-                    "Argument `loss` can only be provided as a dict "
-                    "when the model also returns a dict of outputs. "
-                    f"Received loss={loss}"
+                name = "_".join(flat_output_names[:4])
+                if len(flat_output_names) > 4:  # prevent too long naming string
+                    name += "..."
```

Comment:
Will this causes a collision in `history` when recording individual losses?

Reply:
Potentially.
This case happens when the loss is applied at the root of the model's output. For example if `output=(y1, {'a':y2, 'b': ...})`. How should the (potentially deeply nested) pytree `(y1, {'a':y2, 'b': ...})` be named?
Current implementation:
```
if root is a dict:
  Uses concatenation of the dictionary keys
else:
  Uses concatenation of `model.output_names`
```

The name is checked for a maximum size, and then append `...` if the size is exceeded.

Do you have suggestion for a naming convention? CC @fchollet

Reply:
I removed the truncation.